go/vt/vtgate/vstream_manager.go

```diff
@@ -214,6 +221,58 @@ func (vs *vstream) stream(ctx context.Context) error {
 	return vs.err
 }
 
+func (vs *vstream) sendEvents(ctx context.Context) {
+	var heartbeatDuration time.Duration
+	if vs.heartbeatInterval == 0 {
+		heartbeatDuration = time.Duration(math.MaxInt64) * time.Nanosecond // max time.Duration => 290 years
+	} else {
+		heartbeatDuration = time.Duration(vs.heartbeatInterval) * time.Second
+	}
+
+	timer := time.NewTicker(heartbeatDuration)
+	defer timer.Stop()
+
+	send := func(evs []*binlogdatapb.VEvent) error {
+		if err := vs.send(evs); err != nil {
+			vs.once.Do(func() {
+				vs.err = err
+			})
+			return err
+		}
+		return nil
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			vs.once.Do(func() {
+				vs.err = fmt.Errorf("context canceled")
+			})
+			return
+		case evs := <-vs.eventCh:
+			if err := send(evs); err != nil {
+				vs.once.Do(func() {
+					vs.err = err
+				})
+				return
+			}
+			timer.Reset(heartbeatDuration)
+		case <-timer.C:
+			now := time.Now().UnixNano()
```

Comment:
This could be made simpler!

```suggestion
		case t := <-timer.C:
			now := t.UnixNano()
```